### PR TITLE
GH-43116: [C++][Compute] Mark KeyCompare.CompareColumnsToRowsLarge as large memory test

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

This test consumes more than 4GB memory and causes oom-kill when running with TSAN as reported in #43116 .

### What changes are included in this PR?

Limit its running by marking it as large memory test.

### Are these changes tested?

Change is test.

### Are there any user-facing changes?

None.

* GitHub Issue: #43116

### CODE REVIEW

**Comment by zanmato1984** [comment]
Hi @pitrou , mind to take a look? Thanks.

cc @raulcd

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 1]
```diff
@@ -166,10 +166,9 @@ TEST(KeyCompare, CompareColumnsToRowsTempStackUsage) {
   }
 }
 
-#ifndef ARROW_VALGRIND
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/compare_test.cc:169]: I guess we don't care much about the execution time we are running "large memory tests", even with Valgrind, so removing it.

**Review by raulcd — APPROVED** [review]
LGTM
@felipecrv @bkietz are you ok for me merging this for the release?

**Review by felipecrv — APPROVED** [review]
> LGTM @felipecrv @bkietz are you ok for me merging this for the release?

Sure. Go ahead.

**Pull request merged**